README
# ###################################################
# Copyright (C) 2009 The Unknown Horizons Team
# team@unknown-horizons.org
# This file is part of Unknown Horizons.
#
# Unknown Horizons is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the
# Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
# ###################################################

Unknown Horizons is a 2D realtime strategy simulation with an
emphasis on economy and city building. Expand your small 
settlement to a strong and wealthy colony, collect taxes and 
supply your inhabitants with valuable goods. Increase your 
power with a well balanced economy and with strategic trade 
and diplomacy.

Find more information about Unknown-Horizons (and how to get
it run) on our website.

http://www.unknown-horizons.org/



changelog.txt
CHANGELOG 
Unknown Horizons
-----------------------------------------------------
2009-XX-XX: Milestone 2009.1 (to be released)
* General
  * New: Boadbuilder class with unitcreation, it's possible to build ships
  * New: Support for translating the whole GUI was added, including
         German, French, Portuguese and Norwegian translations, add       
         language choose menu in settings
  * New: Load/save function for maps
  * New: Start work at wildanimals
  * New: Random start position for playership
  * New: add commandline arguments --start-map; --enable-unstable-features;
  * New: Add logging system
  * New: Add debug modul option for commandline --debug-module
  * New: debug messages are now only shown by using ./run_uh.py -d
  * New: Add new tabwidget
  * New: Ship radius when building is now updated on the fly if the ship 
         moves
  * New: Manpage
  * New: Change Style of the mainmenu and all submenus
  * New: Add Developermap with only one small island
  * New: Translations in Portuguese Norwegian German French
  * Change: Hotkeys: set g for grid, b for buildmenu and d for debugger 
  * Fix: Building speed has been improved, lumberjack preview is now   
         almost lagfree.
  * Fix: Loading speed has been improved
  * Fix: Tearing buildings no longer crashes the game
  * Fix: All exceptions when ending a session have been taken care of
  * Fix: Fixed trading bug; selling resources now gives you money
  * Fix: building multiple streets with one click
  * Fix: Main menu doesn't disappear when loading a game

* Graphics and GUI
  * New: Completly new ingame GUI
  * New: Graphics for boat builder
  * New: Graphics for hunter
  * New: Two new variations for inhabitants
  * New: Second model for inhabitants tent
  * Change: Little update on main square
  * Change: Slightly retouched cursor graphics
  * Change: Look of help and creditsmenu
  
* Sound
  * New: Main theme by Tobias Escher

... and a lot more smaller bugfixes.

-----------------------------------------------------

2009-03-08: Milestone 2009.0
* General
  * New: Added a free trader that sells and buys the players goods
  * New: Buildings are now rotatable
  * New: Buy/Sell menu
  * New: Add running costs and an option to decomission buildings to the game
  * New: Started basic work on Gettext support for adding different languages to the game
  * New: Isles get filled randomly with trees on startup
  * New: New Map
  * Change: Changed name from OpenAnno to Unknown Horizons
  * Change: Rebuilt ship inventory
  * Change: Random tree/building on build
  * Change: Lumberjack cuts trees randomly
  * Change: Massive changes in the content directory structure to allow easier implementation of animations
  * Fix: Crash in build mode
  * Fix: Collectors use tents as path
  * Fix: Crash in Destruction mode - mouse drag
  * Fix: Trees being build over existing sreets and trees autotearing them down
  * Fix: Herder goes backwards and his sheeps are all at the same place
  * Fix: Turning menu: picture and building are asynchron
  * Fix: Stock don't increase after building a storage tent
  * Fix: Weater steal in other weater's tent.
  * Fix: Resource Transfer crashes game when out of range
  * Fix: Buildable status only updated on mouse move
  
* Graphics and GUI
  * New: New building main square, where inhabitants get their ressources
  * New: Icons for switching game speed 
  * New: New model for signal fire
  * New: New models for different types of trees
  * Change: Resized buildings and units
  * Change: Added new graphics for units
  * Change: New graphics for main ship and trader
  * Change: Redesign of build menu, added icons of buildable objects
  * Change: New tileset
  * Change: Unified lighting on renderings
  * Change: Updated Buildings: Softened shadows, added some grass, texture work
  * Fix: ship-trademenu and mainmenu are overlapping
  * Fix: Unified lighting in renderings 
  
* Sound
  * New: Implemented build sound
  * New: Added ambient sounds for lumberjack, herder, main square and branch office
  * New: Rolloff function for sound, which fades ambient sounds depending on the position on the map
  * New: New Song Beware the Ship's Kobold!
  * New: Functionality to adjust volume level for music and effects
  * New: Implemented event based speech output
  * New: Added four different sounds to welcome you on the map
  
... and a lot more smaller bugfixes.

Many thanks go to all developers, contributors, testers and supporters.


run_uh.py
#!/usr/bin/env python

# ###################################################
# Copyright (C) 2009 The Unknown Horizons Team
# team@unknown-horizons.org
# This file is part of Unknown Horizons.
#
# Unknown Horizons is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the
# Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
# ###################################################

"""This is the Unknown Horizons launcher, it looks for fife and tries to start the game. If you want to digg
into the game, continue to horizons/main.py. Read all docstrings and get familiar with the functions and
attributes. I will mark all tutorial instructions with 'TUTORIAL:'. Have fun :-)"""

import sys
import os
import gettext
import time
import logging
import logging.config
import logging.handlers
import getopt

def log():
	"""Returns Logger"""
	return logging.getLogger("run_uh")

def get_fife_path():
	"""Returns path to fife engine. Calls sys.exit() if it can't be found."""
	# check if there is a config file (has to be called config.py)
	try:
		import config
		_paths = [config.fife_path]
	except (ImportError, AttributeError):
		_paths = []

	_paths += [ a + '/' + b + '/' + c for a in ('.', '..', '../..') for b in ('.', 'fife', 'FIFE', 'Fife') for c in ('.', 'trunk') ]

	fife_path = None
	for p in _paths:
		if p not in sys.path:
			# check if we are in a fife dir...
			for pe in [ os.path.abspath(p + '/' + a) for a in ('.', 'engine', 'engine/extensions', 'engine/swigwrappers/python') ]:
				if not os.path.exists(pe):
					break
			else:
				fife_path = p

				log().debug("Found FIFE in %s", fife_path)

				#add python paths (<fife>/engine/extensions <fife>/engine/swigwrappers/python)
				for pe in [ os.path.abspath(fife_path + '/' + a) for a in ('engine/extensions', 'engine/swigwrappers/python') ]:
					if os.path.exists(pe):
						sys.path.append(pe)
				os.environ['PYTHONPATH'] = os.path.pathsep.join(\
					os.environ.get('PYTHONPATH', '').split(os.path.pathsep) + \
					[ os.path.abspath(fife_path + '/' + a) for a in \
						('engine/extensions', 'engine/swigwrappers/python') ])

				#add windows paths (<fife>/.)
				os.environ['PATH'] = os.path.pathsep.join( \
					os.environ.get('PATH', '').split(os.path.pathsep) + \
					[ os.path.abspath(fife_path + '/' + a) for a in ('.') ])
				os.path.defpath += os.path.pathsep + \
					os.path.pathsep.join([ os.path.abspath(fife_path + '/' + a) for a in ('.') ])
				break
	else:
		print _('FIFE was not found.')
		exit()
	return fife_path

def find_FIFE():
	"""Inserts path to fife engine to $LD_LIBRARY_PATH (environment variable).
	If it's already there, the function will return, else
	it will restart uh with correct $LD_LIBRARY_PATH. """
	fife_path = get_fife_path() # terminates program if fife can't be found

	os.environ['LD_LIBRARY_PATH'] = os.path.pathsep.join( \
		[ os.path.abspath(fife_path + '/' + a) for  \
			a in ('ext/minizip', 'ext/install/lib') ] + \
		  (os.environ['LD_LIBRARY_PATH'].split(os.path.pathsep) if \
			 os.environ.has_key('LD_LIBRARY_PATH') else []))

	log().debug("Restarting with proper LD_LIBRARY_PATH...")
	log().debug("LD_LIBRARY_PATH: %s", os.environ['LD_LIBRARY_PATH'])
	log().debug("PATH: %s", os.environ['PATH'])
	log().debug("PYTHONPATH %s", os.environ['PYTHONPATH'])

	# assemble args (python run_uh.py ..)
	args = [sys.executable] + sys.argv + [ "--fife-in-library-path"]
	#we are already in Unknown Horizons root, so just exec local executeable
	args[1] = os.path.split(os.path.realpath(args[1]))[1]
	os.execvp(args[0], args)

def print_help():
	print _("Unknown Horizons usage:")
	print "run_uh.py [-d] [-h]"
	print ""
	print _("Options:")
	print "-d --debug   - ", _("Debug, enables debug output, useful for testing.")
	print "-h --help    - ", _("Print this help message.")
	print ""
	print "--start-dev-map   - ", _("Starts the development map without displaying the main menu")
	print "                    ", _("Useful for testing during development")
	print "--start-map <map> - ", _("Starts <map>. <map> is the filename of the map, without '.sqlite'")
	print "--load-map <save> - ", _("Loads a saved game. Specify the savegamename.")
	print ""
	print _("Debugging options:")
	print "--debug-module <module> -", _("Enable logging for a certain logging module.")
	print ""
	print _("Have fun playing, and if you do, help us developing!")


def find_uh_position():
	"""Returns path, where uh is located"""
	first_guess = os.path.split( os.path.realpath( sys.argv[0]) )[0]
	if os.path.exists('%s/content' % first_guess):
		return first_guess
	else:
		positions = ['/usr/share/games',
			     '/usr/share',
			     '/usr/local/share/games',
			     '/usr/local/share',
			     ]

		for i in positions:
			if os.path.exists('%s/unknown-horizons' % i):
				return '%s/unknown-horizons' % i


if __name__ == '__main__':


	#chdir to Unknown Horizons root
	os.chdir( find_uh_position() )
	logging.config.fileConfig('content/logging.conf')
	gettext.install("unknownhorizons", "po", unicode=1)

	# parse arguments
	try:
		opts, args = getopt.getopt(sys.argv[1:], "hd", \
						   ["help", "debug", "fife-in-library-path", "start-dev-map", \
							    "start-map=", "enable-unstable-features", "debug-module=", \
									"load-map="])
	except getopt.GetoptError, err:
		print str(err)
		print_help()
		exit(1)

	fife_in_library_path = False

	command_line_arguments = { \
		         "start_dev_map": False, \
					   "start_map": None, \
						 "load_map": None,
					   "unstable_features": False, \
					   "debug": False }

	# apply arguments
	for o, a in opts:
		if o in ("-h", "--help"):
			print_help()
			exit(1)
		elif o in ("-d", "--debug"):
			logging.getLogger().setLevel(logging.DEBUG)
			command_line_arguments['debug'] = True
			# init the logfile handler with a dynamic filename
			logfilename = "unkown-horizons-"+time.strftime("%y-%m-%d_%H-%M-%S")+".log"
			file_handler = logging.FileHandler(logfilename, 'w')
			logging.getLogger().addHandler(file_handler)
		elif o == "--fife-in-library-path":
			# this is currently only for internal use, therefore not in the help message
			fife_in_library_path = True
		elif o == "--start-dev-map":
			# automatically starts development map
			command_line_arguments['start_dev_map'] = True
		elif o == "--start-map":
			# start map selected by commandline arg
			command_line_arguments['start_map'] = a
		elif o == "--load-map":
			# load map selected by commandline arg
			command_line_arguments['load_map'] = a
		elif o == "--enable-unstable-features":
			command_line_arguments["unstable_features"] = True
		elif o == "--debug-module":
			# enable logging for this module
			logging.getLogger(a).setLevel(logging.DEBUG)


	#find fife and setup search paths, if it can't be imported yet
	try:
		import fife
	except ImportError, e:
		if fife_in_library_path:
			# fife should already be in LD_LIBRARY_PATH
			print 'Failed to load fife:', e
			exit(1)
		log().debug('Searching for FIFE')
		find_FIFE()

	#print _("Launching Unknown Horizons")

	#for some external libraries distributed with unknownhorizons
	sys.path.append('horizons/ext')

	#start unknownhorizons
	import horizons.main
	horizons.main.start(command_line_arguments)
